### ISSUE TITLE
Feature request: Drag by mouse within a playlist to change the order

### ISSUE THREAD

**Issue by Gero5** [mpogue2/SquareDesk#1074]
(V 1.0.7)
problem: songs from the song table can only be added to the bottom of the playlist. to change the sequence within the playlist you need several "move up/down" actions. that devours a lot of time when needed for several songs.

new feature: add a drag symbol to the  playlist entry. klick and hold it and drag the song to the right position in the list.
![drag](https://github.com/mpogue2/SquareDesk/assets/128936259/0e43efae-2152-400d-989d-684ad8e51510)

**Comment by mpogue2** [comment]
That's a good idea!  Drag and drop is rather difficult to implement, so it has not been worked on yet.  But, it should be possible (just not simple to do).

**Comment by phoef** [comment]
Oh great! When I read that you are looking into using AI support for complex challenges, I immediately thought about the idea of adjusting playlists with drag & drop. I searched for the issue — and saw that it’s already tagged with “Claude?” :-) I would be really happy if something could be implemented in that direction.

**Comment by mpogue2** [comment]
@phoef I am experimenting with AI for coding, with some great successes and some failures.

1) Claude did NOT succeed at this task, although it had the right general idea.  Claude did a FANTASTIC job on a couple of other items I had it do.
2) I couldn't figure out how to get Gemini Pro 2.5 to work on the problem at all....it just hung, so I gave up.
3) I was able to get Copilot (Microsoft) to do about 90% of it, and then I had to do the other 10%.  I still have a few bugs around coloring of playlist items that are dragged in, but the basics of drag and drop between the three types of lists seems to be working and stable.

The three types of lists right now:
- track filters, like "patter"
- regular playlists
- the main songTable

Not all of these drag/drop combinations are valid, for example you can't reorder a track filter or the songTable.  But, you CAN now drag within a playlist, between playlists, and between the songTable and a playlist.

Commit: 8855c3172cc2a4063f96bdfac5bf89266ff4da4f

**Comment by mpogue2** [comment]
Wow, #11 was entered in 2017!  Finally implemented in 2025. :-)

**Comment by phoef** [comment]
Cool! The combinations you mentioned are exactly what I had in mind. Thank you!

**Comment by phoef** [comment]
In a podcast I heard about https://www.cursor.com as AI Code Editor ??? Perhaps you already tested it? Perhaps it's worth a try?

**Comment by Gero5** [comment]
Cool, if this works. Can't wait to test it out.
And now I don't have to wonder anymore who or what "Claude?" is....

Claude, Code Pilot, Gemini,  Cursor, ChaptGPT, Apple AI
in Germany we say "Too many (AI) cooks spoil the broth.", but if it works for coding I do not complain. ;-)

**Comment by mpogue2** [comment]
I think coding is going to be a big use of Ai….its like having an awesome junior programmer working for me, who knows how to research questions, but who also makes a lot of dumb mistakes. So far I have tried Cursor, Windsurf, VS Code, Claude Code, Roo Code, AI Studio, Copilot, and Claude Desktop.

And, the winner has been Claude Desktop + Sonnet + Desktop Commander.  When it can solve my problem, it’s like magic!  I describe what I want in detail, and it does all the grunt work. And it’s super fast. When it can’t solve the problem, it gets stuck, but doesn’t know it’s stuck (I have to tell it in detail what it’s done wrong).  This combination did solve a couple of problems that *I* was stuck on, saving me huge amounts of time.

**Comment by mpogue2** [comment]
This feature will be in the next release (hopefully this weekend).

Note: I did not implement the 12-dot icon part.  You can start dragging an item from **anywhere on that item**.  If you see a short version of the item's title moving with your cursor, then you've grabbed it.  If you see a bright green line somewhere, that's where you can drop what you've grabbed.  No green line, can't drop there.

**Comment by mpogue2** [comment]
LOL -- I just realized that you can drag items out of SquareDesk into a text editor, and the text editor will show you the very messy string that SquareDesk uses internally for drag and drop, that contains all the information about those songs.  

I'm not sure that this is a useful, or even a Good Thing to be able to do... :-)

**Comment by Gero5** [comment]
>No green line, can't drop there.
The indicator is a green circle with a white plus sign here. 

I can drag an item from the song table to the palette. But it is added to the bottom only.
I can drag an item from one palette to another palette. But it is added to the bottom only even when I drop it at the top.
But I cannot change the order within the same playlist using drag and drop. Like drag # 3 to position # 10. although the plus sign is displayed.
@Claude: Is it possible to implement that as well?

**Comment by mpogue2** [comment]
@Gero5 What version are you on?
- I never see a green circle with a plus sign here. should see the first 20 characters of the song name.
- playlist drag drop works every time for me.

Hmmm...This leads me to believe that you are one version behind.  Make sure you are using v1.1.1 ....

**Comment by Gero5** [comment]
v1.1.0

![Image](https://github.com/user-attachments/assets/3340dc0f-d378-48dd-8811-b32a3928aefa)

**Comment by Gero5** [comment]
With v 1.1.1 I have the green circle with the plus sign AND the green line. :-)
Drag and drop works within playlists on the palette, between the palettes, but not on a playlist displayed in the song table.
Items from the song table work one way out to palettes only.

**Comment by mpogue2** [comment]
Yes, hmmm I'll have to think about that.  Definitely an oversight on my part! The song table is currently read-only by design, ands what's there is never editable. I did not think that there might be something editable there.

**Comment by mpogue2** [comment]
Btw, the pace is going to slow down soon...I pushed hard to make changes for 1.1.1, and I need a break! :-)

**Comment by Gero5** [comment]
Changes to the playlist are stored in the palette only but not in the song table. With the same playlist displayed in the palette and the song table simultaneously only the palette shows the changes made by drag and drop. This should be synchronized or  the changes should be saved and updated in all the tables where it is displayed. song table is the only other I guess.

> Btw, the pace is going to slow down soon...I pushed hard to make changes for 1.1.1, and I need a break! :-)

You deserve it. Thank you for keeping up the good work.

**Comment by mpogue2** [comment]
See #1404...

**Comment by phoef** [comment]
> Btw, the pace is going to slow down soon...I pushed hard to make changes for 1.1.1, and I need a break! :-)

Totally understandable – you really earned that break! The recent updates and upgrades have been amazing – both in terms of content and how quickly you reacted. Thanks a lot for your great work – it’s clear how much passion you put into the project. 🙌

**Comment by Gero5** [comment]
> The song table is currently read-only by design, ands what's there is never editable.

So you have to create an editable mode for the song table when it displays a playlist and not a song selection.

But maybe it is not that dramatic if it stays in read only. We can organize the playlist in the palette. The user's gut feeling (or just mine?) is that what works in the palette should also work in the playlist shown in the song table.

**Comment by mpogue2** [comment]
Yeah, I think there is an expectation there.

In the short term, Aat minimum, I would want to indicate somehow that the songTable is Read-Only.  It already does not act as a drop-enabled table would (no green line), but maybe the Read-Only nature of the songTable needs to be more explicit.

As #1404 suggests, there's already a relationship between palette slot and song table that is not implemented, namely when the palette slot is edited, the songTable showing that same playlist should be updated.  And that would probably need to be two-way, if the song table contents were editable.  And, you'd probably need to be able to edit the songTable by dragging items up and down, too (if it was showing a Playlist, but not if it was showing an Apple Music playlist or a Track Filter), just like a playlist slot.  

Essentially the mental model could be, as you suggest, that these are just 4 views of things, and 3 of them happen to be smaller than the 4th.  And, Track Filter things and Apple Music things are not editable, while Playlist things are.

I'll need to think about this some more...My mental model always has been that songTable is a read-only view of things, and the palette slots are views that are sometimes read/write if they contain SquareDesk playlists.

Much to think about...

**Comment by Gero5** [comment]
note that consequently the context menu in the song table has to be the same as in the palette for playlist items. 
Move up/down, remove cannot be done in the song table at the moment. And what about "Save..."? There has to be an option for the edited playlist in the song table as well.
